### ISSUE TITLE
how can I add a relative path to api-docs/  ?

### ISSUE THREAD

**Issue by bluelovecute** [springfox/springfox#569]
for example ： i add rest/ to the basePath but it did not work

**Comment by dilipkrish** [comment]
@bluelovecute Could you elaborate yr use case? Just changing basePath is not going to make your application create RequestMappings at that endpoint.

**Comment by bluelovecute** [comment]
First, please forgive me for my poor spoken English.
I use the v0.9.4 and  config it by spring.xml 
```java
   CustomSpringSwaggerConfig.class
   @Bean
    public SwaggerSpringMvcPlugin customImplementation(ServletContext servletContext){
        return new SwaggerSpringMvcPlugin(this.springSwaggerConfig)
                .apiInfo(apiInfo())
                .includePatterns("/rest/.*")
                .pathProvider(swaggerPathProvider(servletContext))
                ;
    }
```

my project  need to add "/rest/"  to support RestController, and I need to add "/rest/"  in front of the api-docs/ to escape the interceptor,or else add a suffix ".html" (but it not design to use for RestController) ,
 only change the *pathProvider* can't change the apiPath? I have tried to set SwaggerConfig.calss 's property apiPath to "/rest/api-docs" too, but it doesn't make sense. 

the result:
```javascript
 Web User Controller at location: /api-docs http://localhost:8080/easydm-mobile/rest/api-docs/default/web-user-controller

the url doesn't work 
http://localhost:8080/easydm-mobile/rest/api-docs  --- not work
http://localhost:8080/easydm-mobile/api-docs.htm  ---work 
```
Is it because of DefaultSwaggerController 's property DOCUMENTATION_BASE_PATH set a default value "/api-docs" ? and how can I change it? 
how can I change the location?
Thanks!

**Comment by coldfusor** [comment]
I think I have the same basic question as @bluelovecute.

Right now I have an application with api-docs at `http://server:port/foo/api-docs` and endpoints at `http://server:port/foo/endpointname`

My web.xml includes the following dispatching logic:

    <servlet>
        <servlet-name>DispatcherServlet</servlet-name>
        <servlet-class>org.springframework.web.servlet.DispatcherServlet</servlet-class>
        <init-param>
            <param-name>contextClass</param-name>
            <param-value>org.springframework.web.context.support.AnnotationConfigWebApplicationContext</param-value>
        </init-param>
    </servlet>
    <servlet-mapping>
        <servlet-name>DispatcherServlet</servlet-name>
        <url-pattern>/*</url-pattern>
    </servlet-mapping>

I need to move my api-docs to `http://server:port/bar/api-docs` while keeping the application endpoints where they're at.

**Comment by adrianbk** [comment]
@coldfusor @bluelovecute have you resolved this? 

https://github.com/martypitt/swagger-springmvc/blob/master/swagger-springmvc/src/main/java/com/mangofactory/swagger/paths/SwaggerPathProvider.java#L23

**Comment by coldfusor** [comment]
From what I understand, setting the apiResourcePrefix only changes what Swagger documents as the endpoints' base path. What I'd like to do is change the path where api-docs itself is hosted.

For endpoints themselves, I can always change a request mapping path from `/endpointname` to `/foo/endpointname`, but it seems that api-docs is always hosted at the lowest level of the servlet, e.g. `/api-docs`. I want to push it in to `/foo/api-docs` and leave the other paths unchanged. Right now the only way I can see is to change my DispatcherServlet pattern from `/*` to `/foo/*` which would affect all of the endpoints as well (which isn't compatible with the path structure my application needs).

If [DOCUMENTATION_BASE_PATH](https://github.com/martypitt/swagger-springmvc/blob/master/swagger-springmvc/src/main/java/com/mangofactory/swagger/controllers/DefaultSwaggerController.java#L22) were made editable, would changing that variable change the path to api-docs? I think it would, based on the request mappings in DefaultSwaggerController. If so, that's the change I would like to see; I could change my DOCUMENTATION_BASE_PATH from `/api-docs` to `/foo/api-docs`.

Side note: I edited the paths in my previous comment to better reflect my setup.

**Comment by dilipkrish** [comment]
@coldfusor sounds like [this is the solution](http://stackoverflow.com/questions/17128463/overriding-requestmapping-on-springmvc-controller). Would you like to try and submit a pr for this?

**Comment by coldfusor** [comment]
By

**Comment by coldfusor** [comment]
Do you mean using a properties file to set the api-docs base path? That doesn't line up with how the rest of the application is configured with builder objects.

**Comment by dilipkrish** [comment]
@coldfusor :confused: not sure what builders you're referring to

**Comment by adrianbk** [comment]
Let's not introduce properties files if at all possible..

@coldfusor you can have multiple servlet mappings to the same dispatcher servlet:
```
 <servlet-mapping>
    <servlet-name>dispatcher</servlet-name>
    <url-pattern>/foo/*</url-pattern>
  </servlet-mapping>
  <servlet-mapping>
    <servlet-name>dispatcher</servlet-name>
    <url-pattern>*.html</url-pattern>
  </servlet-mapping>

```

You could also extend DefaultSwaggerController and effectively override the paths.

**Comment by coldfusor** [comment]
@dilipkrish For example, building a SwaggerSpringMvcPlugin like

    @Bean
    public SwaggerSpringMvcPlugin swaggerSpringMvcPlugin() {
        return new SwaggerSpringMvcPlugin(this.springSwaggerConfig)
                .apiInfo(apiInfo())
                .apiVersion("1.0")
                .pathProvider(customSwaggerPathProvider());
    }

or some things are configured by extending them, e.g. I have

    private SwaggerPathProvider customSwaggerPathProvider() {
        return new SwaggerPathProvider(){
            @Override
            protected String applicationPath() {
                return "/foo/bar";
            }
            
            @Override
            protected String getDocumentationPath() {
                return "/";
            }
        };
    }

**Comment by coldfusor** [comment]
@adrianbk I'm certainly not a fan of properties files in this situation. :)

If I do multiple servlet mappings, how do I restrict resources so that they're only accessed by one of the patterns? E.g. if I have mappings for `/foo/*` and `/bar/*` and I want to have `/foo/api-docs` but not `/bar/api-docs`. I'm pretty new to Java servlets so maybe this is a trivial problem, but I've Googled for an explanation before with no luck.

Extending DefaultSwaggerController could be a solution, but various places in swagger-springmvc specifically import DefaultSwaggerController, so I'd have to reconfigure a lot of the code. I'd much rather just be able to set my own DOCUMENTATION_BASE_PATH.

**Comment by dilipkrish** [comment]
@adrianbk wasn't suggesting adding property files. 

However, It seems reasonable to me to have be extensible. As outlined in the SO post, it could be property expressions can be defaulted. 

For e.g. 
```java
@RequestMapping("${documentation.root.path:/api-docs}")
```

That way property files are not mandated, but still support extensibility. thots?

**Comment by coldfusor** [comment]
Configuring DOCUMENTATION_BASE_PATH with a property, even one with a default value, still breaks with the configuration style of other variables in the project. For consistency's sake I would like to add a configuration method to SwaggerSpringMvcPlugin, perhaps called documentationBasePath(), so the setup would look like:

    @Bean
    public SwaggerSpringMvcPlugin swaggerSpringMvcPlugin() {
        return new SwaggerSpringMvcPlugin(this.springSwaggerConfig)
                .documentationBasePath("/foo/bar"); // new method
    }

**Comment by coldfusor** [comment]
Any objections to that scheme? If not, I'll start writing some code.

**Comment by dilipkrish** [comment]
@coldfusor 

Couple of things,

I'm not sure how you'd be able to override that at runtime. To my knowledge RequestMapping is a source level construct and requires a constant value or a spring expression.

See this [spring boot example](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/BasicErrorController.java#L63-L66). 

Secondly, the SwaggerSpringMvcPlugin is a class to influence the documentation generation. Its not to  configure the controller.

Lastly and more importantly, we're about to merge our [2.0 effort](https://github.com/springfox/springfox/tree/refactor/538/spike-to-seperate-models-and-dtos) back to mainline. So writing a solution based on 1.0 codebase may never be released :open_mouth: 

@adrianbk want to weigh in?

**Comment by dilipkrish** [comment]
@coldfusor this should now be fixed in 2.0. See [here](https://github.com/springfox/springfox/blob/master/transitioning-to-v2.md#customizing-the-swagger-endpoints) for more information.

**Comment by coldfusor** [comment]
@dilipkrish Excellent, thank you! I'm looking forward to the 2.0 release.
